rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/DynamicCapabilities.java

```diff
@@ -0,0 +1,205 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.rascalmpl.vscode.lsp.parametric.capabilities;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.eclipse.lsp4j.ClientCapabilities;
+import org.eclipse.lsp4j.Registration;
+import org.eclipse.lsp4j.RegistrationParams;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.Unregistration;
+import org.eclipse.lsp4j.UnregistrationParams;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.rascalmpl.vscode.lsp.parametric.ILanguageContributions;
+import org.rascalmpl.vscode.lsp.parametric.LanguageContributionsMultiplexer;
+import org.rascalmpl.vscode.lsp.util.concurrent.CompletableFutureUtils;
+
+/**
+ * Takes care of (un)registering capabilities dynamically.
+ * Receives a list of supported capabiltities ({@link AbstractDynamicCapability}) and notifies the client when capabilities change.
+ */
+public class DynamicCapabilities {
+
+    private static final Logger logger = LogManager.getLogger(DynamicCapabilities.class);
+
+    private final LanguageClient client;
+    private final Executor exec;
+    private final List<AbstractDynamicCapability<?>> supportedCapabilities;
+    private final Set<AbstractDynamicCapability<?>> staticCapabilities;
+    private final Map<String, Registration> currentRegistrations = new ConcurrentHashMap<>();
+
+    private @MonotonicNonNull ClientCapabilities clientCapabilities;
+    private @MonotonicNonNull ServerCapabilities serverCapabilities;
+
+    public DynamicCapabilities(LanguageClient client, Executor exec, List<AbstractDynamicCapability<?>> supportedCapabilities) {
+        this.client = client;
+        this.exec = exec;
+        this.supportedCapabilities = supportedCapabilities;
+        this.staticCapabilities = new HashSet<>();
+    }
+
+    /**
+     * Determines whether the server should register capabiltities statically now instead of dynamically later.
+     * @param clientCapabilities The capabilities of the client, which determine whether dynamic registration is supported at all.
+     * @param serverCapabilities The server capabilities to modify.
+     */
+    public void setStaticCapabilities(ClientCapabilities clientCapabilities, final ServerCapabilities serverCapabilities) {
```

Comment:
I think the name is not correct, as it also sets the client dynamic capabilities?

Reply:
Changed these functions to improve the responsibilities and timing.